### ISSUE TITLE
Option to ignore versions in pom.xml without need for Rules.xml

### ISSUE THREAD

**Issue by simdevmon** [mojohaus/versions-maven-plugin#258]
I think it would be convenient if I could specify versions to ignore directly in `pom.xml` under the `configuration` tag. Then I would not have the need to create an extra `Rules.xml` file, but could simply specify everything directly in the pom:

```
<ignoreVersions>
    <ignoreVersion type="regex">.*-beta.</ignoreVersion>
    <ignoreVersion type="regex">.*_ALPHA</ignoreVersion>
</ignoreVersions>
```

**Comment by sbrunov** [comment]
Hi, everyone!

Completely agree with the previous authors.

The requested functionality is quite essential, so it would be great to have it.

Best regards,  
Sergey Brunov.

**Comment by svbrunov** [comment]
Dear maintainers,

Please, consider re-opening this GitHub issue.

Best regards,
Sergey Vyacheslavovich Brunov.

**Comment by garretwilson** [comment]
How can I vote for this? I have a bounty running [on Stack Overflow](https://stackoverflow.com/q/72416739) to try to find a workaround. Most plugins don't make me create an entire separate project just to configure them and have there configurations be present in child projects. Why should this one be any different?

**Comment by garretwilson** [comment]
Could someone give an update on whether there is any thought of fixing this soon? We're about to release a new version of our POM and it would be good to know if this is at least scheduled tentatively or if there isn't any hope for it being addressed.

**Comment by ajarmoniuk** [comment]
It looks like a feature like this has recently been implemented by #369. Please see #318 where it's possible to provide inclusion and exclusion filters for determining which dependency patterns will be considered. Thanks to that, you can rule out patterns such as `.*-beta.` or `.*_ALPHA`, albeit not using regexp, but simple asterisk wildcards.

**Comment by garretwilson** [comment]
@ajarmoniuk , can you clarify? Are these inclusions and exclusions with wildcards something that I can put in the POM? (The examples at #369 only show JVM property usage on the command line.) I don't see this mentioned in the [`display-dependency-updates` documentation](display-dependency-updates).

**Comment by garretwilson** [comment]
Woohoo! @ajarmoniuk you don't know how excited I am to try this out.

> I don't know why plugins and pluginManagement sections have not got filtering

Yeah that's the next thing I thought of to ask you. 😄 I think the dependencies section is the most important, but I'll bet we'll need them in the other sections too.

OK, I'm 🤞 waiting for the release!

**Comment by ajarmoniuk** [comment]
Yes, you seem to understand it correctly.

Well, it's quite complex actually.

Include means only show the dependencies satisfying the filter.
Exclude means show everything that is included (so, really _everything_ if there's no include filter) **except** the ones matching the _exclude_ filter.

**Comment by ajarmoniuk** [comment]
So, I've just added the missing plugin- and plugin management filtering which works likewise. I really doubt it will land into today's release though.

**Comment by garretwilson** [comment]
> So, I've just added the missing plugin- and plugin management filtering which works likewise. I really doubt it will land into today's release though.

@ajarmoniuk , do you happen to know if the plugin-related exclusion configuration made it into v2.12.0?

Secondly, an odd thing: for some reason `versions-maven-plugin` in my POM doesn't seem to realize that itself is out of date; I'm using v2.11.0, but v2.12.0 is available. But `mvn versions:display-plugin-updates` is only listing _other_ plugins_ as being out of date. Is there some sort of lag before `versions-maven-plugin` will see a new version? Doesn't it go off Maven Central?

**Comment by slawekjaranowski** [comment]
@garretwilson it was published to central ...

https://search.maven.org/artifact/org.codehaus.mojo/versions-maven-plugin

**Comment by maffe** [comment]
I think your local Maven installation queries Central only once a day for known artifacts. Try deleting the data in your local repository.

**Comment by ajarmoniuk** [comment]
Try using -U with the maven command.

**Comment by slawekjaranowski** [comment]
1. look at release notes https://github.com/mojohaus/versions-maven-plugin/releases/tag/versions-maven-plugin-2.12.0 - it was  included

**Comment by garretwilson** [comment]
Thanks for info and tips.

**Comment by garretwilson** [comment]
> Try using -U with the maven command.

The `-U` switch (in `mvn -U versions:display-plugin-updates`) fixed the problem; it made Maven go and find the new v2.12.0. I guess it was caching something.

**Comment by garretwilson** [comment]
Unfortunately the milestone version dependency exclusion still isn't working for me. I'm starting with this POM:

https://github.com/globalmentor/globalmentor-root/blob/a78abf59b9609fcbd6bf4f41bea0ea35025067f2/pom.xml

Then in my `<build><pluginManagement>` section I set the configuration like this:

```xml
<plugin>
 <groupId>org.codehaus.mojo</groupId>
  <artifactId>versions-maven-plugin</artifactId>
  <version>2.12.0</version>
  <executions>
    <execution>
      <goals>
        <goal>display-plugin-updates</goal>
      </goals>
      <configuration>
        <pluginDependencyExcludes>*:*:*-M*</pluginDependencyExcludes>
        <pluginManagementDependencyExcludes>*:*:*-M*</pluginManagementDependencyExcludes>
      </configuration>
    </execution>
  </executions>
</plugin>
```

Then from the command line I issue:

```
mvn versions:display-plugin-updates
```

I still get this:

```
[INFO] --- versions-maven-plugin:2.12.0:display-plugin-updates (default-cli) @ globalmentor-root ---
[INFO]
[INFO] The following plugin updates are available:
[INFO]   maven-failsafe-plugin .......................... 2.22.2 -> 3.0.0-M7
[INFO]   maven-release-plugin ............................ 2.5.3 -> 3.0.0-M6
[INFO]   maven-site-plugin .............................. 3.12.1 -> 4.0.0-M3
[INFO]   maven-surefire-plugin .......................... 2.22.2 -> 3.0.0-M7
```

Why isn't it excluding the `*-M*` versions as I asked it to?

It looks like it isn't picking up the exclusions at all. If I skip discussing executions and try to to exclude everything for all goals:

```xml
<plugin>
 <groupId>org.codehaus.mojo</groupId>
  <artifactId>versions-maven-plugin</artifactId>
  <version>2.12.0</version>
  <configuration>
    <pluginDependencyExcludes>*</pluginDependencyExcludes>
    <pluginManagementDependencyExcludes>*</pluginManagementDependencyExcludes>
  </configuration>
</plugin>
```

It still doesn't exclude anything. And if I do a `mvn help:effective-pom`, the effective POM doesn't show any other configuration. Why isn't this configuration taking effect?

**Comment by ajarmoniuk** [comment]
Well, you're using the wrong goal. Those changes were implemented for `display-dependency-updates`, and not for `display-plugin-updates` yet.

Display dependency updates wrt plugins and plugin management display plugin dependency updates of plugins with dependencies, like so:

```xml
  <build>
    <plugins>
      <plugin>
        <groupId>localhost</groupId>
        <artifactId>dummy-maven-plugin</artifactId>
        <version>1.0</version>
        <dependencies>
          <dependency>
            <groupId>localhost</groupId>
            <artifactId>dummy-api</artifactId>
            <version>1.0</version>
          </dependency>
        </dependencies>
      </plugin>
    </plugins>
    <pluginManagement>
      <plugins>
        <plugin>
          <groupId>localhost</groupId>
          <artifactId>dummy-maven-plugin</artifactId>
          <version>1.0</version>
          <dependencies>
            <dependency>
              <groupId>localhost</groupId>
              <artifactId>dummy-impl</artifactId>
              <version>1.0</version>
            </dependency>
            <dependency>
              <groupId>localhost</groupId>
              <artifactId>dummy-parent2</artifactId>
              <version>1.0</version>
            </dependency>
          </dependencies>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
```

As for the new ask, could probably do that as well, could you please create another issue?

**Comment by garretwilson** [comment]
> Well, you're using the wrong goal. Those changes were implemented for `display-dependency-updates`, and not for `display-plugin-updates` yet.

Wait, what?!?

The purpose of `display-dependency-updates` is to check if there are newer dependencies. The purpose of `display-plugin-updates` is to check to see if there are new plugins. Why on earth did you add `pluginDependencyExcludes` to `display-plugin-updates`, which _doesn't even check plugins_?????

So what does

> As for the new ask …

There is no new ask. This ticket is about the general exclusion feature of Versions Maven Plugin. The plugin always had an `<ignoreVersions>` exclusion capability. This request was to allow that same configuration to be configured in the POM like almost every other plugin out there, instead of some external file that doesn't get inherited down the POM hierarchy.

Instead of what this ticket was for, it seems that some other, _additional_,  competing bu incompatible exclusion capability was implemented, but only halfway, in an inconsistent manner, and not for all the check-for-update goals. I really don't know what to say.

**Comment by arjantijms** [comment]
In my case Maven keeps bugging me about the following update:

```
[INFO] --- versions:2.14.2:display-dependency-updates (specific) @ ldapbp ---
[INFO] The following dependencies in Dependency Management have newer versions:
[INFO]   antlr:antlr ........................................ 2.7.7 -> 20030911
```

So I excluded this as per the documentation, I think:

```
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>versions-maven-plugin</artifactId>
                <version>2.14.2</version>
                <executions>
                    <execution>
                        <id>specific</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>display-dependency-updates</goal>
                        </goals>
                        <configuration>
                            <dependencyExcludes>antlr:antlr:20030911</dependencyExcludes>
                            <dependencyManagementExcludes>antlr:antlr:20030911</dependencyManagementExcludes>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

And I invoke it via;

```
mvn versions:display-dependency-updates@specific
```

But it keeps showing the `antlr:antlr:20030911` update. What do I need to do to actually exclude it? Am I doing something wrong, or is this a bug in the plug-in?

**Comment by ajarmoniuk** [comment]
> Am I doing something wrong

Yes.

You're matching your dependency against a version which you are not using in your pom. You should pin it with the one you're using, e.g.
```
antlr
antlr:antlr
antlr:antlr:*
antlr:antlr:2.7.7
```

etc.

**Comment by arjantijms** [comment]
With:

```xml
<configuration>
  <dependencyManagementExcludes>antlr</dependencyManagementExcludes>
</configuration>
```

It will not tell me when 2.7.8 comes out, does it?

**Comment by ajarmoniuk** [comment]
Oh, snap. You mean filtering on the output dependencies. I guess that isn't working indeed. I wonder if it ever worked. In any case, it should be quite easy to fix.

**Comment by ajarmoniuk** [comment]
Created #921

**Comment by ajarmoniuk** [comment]
Already implemented.

Please see https://www.mojohaus.org/versions/versions-maven-plugin/display-dependency-updates-mojo.html#ignoredVersions

**Comment by arjantijms** [comment]
>  You mean filtering on the output dependencies.

Indeed, not sure why one would ever want to filter dependencies you already have, for getting suggestions about dependencies you don't have. The example above specifically is about excluding milestone versions:

```xml
<dependencyExcludes>*:*:*-M*</dependencyExcludes>
```

That's not about the user's pom already being full of `*-M*` versions, but the user not wanting to get suggestions to update to `*-M*` versions. The user above literally says so:

```
And after adding these exclusions, the plugin would no longer say that org.apache.maven.doxia:doxia-core had a newer version, as the newer version 2.0.0-M3 is a milestone matching the excludes. But if a version 1.11.2 came available, it would notify me of that.
```

So the user want's to exclude the new version `2.0.0-M3`; not wanting to have that as suggestion and therefor not update to that. That seemed to be the entire purpose of the option. But now it seems that instead of ignoring the new version, it will ignore `2.0.0-M3` if it's in the pom, and then never suggest say `1.11.2`. But that's entirely the opposite of what people want. You do not want `2.0.0-M3` in the pom, and you do want `1.11.2` to be suggested.


Now `<ignoredVersions>` does work, and it seems to be good enough for my particular use case.

But in general, unless I'm *really* missing something, it looks like `dependencyExcludes` is implemented wrongly? https://github.com/mojohaus/versions/issues/921 is still correct then. As you'd still want to say to exclude suggestions for `foo:bar:*-M*` and `kaz:zak:*-RC*`, right?

**Comment by ajarmoniuk** [comment]
Glad that it worked for you.

> https://github.com/mojohaus/versions/issues/921 is still correct then. As you'd still want to say to exclude suggestions for foo:bar:*-M* and kaz:zak:*-RC*, right?

Well, yes and no. The original intention of this "feature" was to filter the input dependencies. I agree it makes little sense to go over the version there. Maybe a new feature toggle to change the behaviour? I can add one later.

@slawekjaranowski what do you think?

**Comment by garretwilson** [comment]
I'm circling back around to finally test this issue and #684.

* Just to make sure I'm understanding this: what was indicated in this ticket was not actually a replacement for the `rules.xml` file `<ignoreVersions>` section?
* If I want to ignore certain dependency versions when checking for updates, I should not use the facility in this ticket but instead that described at [Using the ruleSet element in the POM](https://www.mojohaus.org/versions/versions-maven-plugin/version-rules.html#using-the-ruleset-element-in-the-pom).
* If I set up the rules set in the POM, this will apply to plugin dependency versions as well (#684).
* There appears also to be a [`<ignoredVersions>`](https://www.mojohaus.org/versions/versions-maven-plugin/display-dependency-updates-mojo.html#ignoredVersions) configuration option now as well that does the same thing? (I didn't see this in the documentation anywhere.)
* Or I can use the user property `maven.version.ignore` to do the same thing.

Do I understand the current situation correctly?
